lib/handlers/token-handler.ts

```diff
@@ -0,0 +1,331 @@
+  ClientCredentialsGrantType,
+  PasswordGrantType,
+  RefreshTokenGrantType,
+} from '../grant-types';
+import { Client, Model } from '../interfaces';
+import { TokenModel } from '../models';
+import { Request } from '../request';
+import { Response } from '../response';
+import { BearerTokenType } from '../token-types';
+import { hasOwnProperty } from '../utils/fn';
+import * as is from '../validator/is';
+
+/**
+ * Grant types.
+ */
+
+const grantTypes = {
+  authorization_code: AuthorizationCodeGrantType,
+  client_credentials: ClientCredentialsGrantType,
+  password: PasswordGrantType,
+  refresh_token: RefreshTokenGrantType,
+};
+export class TokenHandler {
+  accessTokenLifetime: any;
+  grantTypes: { [key: string]: any };
+  model: Model;
+  refreshTokenLifetime: number;
+  allowExtendedTokenAttributes: boolean;
+  requireClientAuthentication: any;
+  alwaysIssueNewRefreshToken: boolean;
+  constructor(options: any = {}) {
+    if (!options.accessTokenLifetime) {
+      throw new InvalidArgumentError(
+        'Missing parameter: `accessTokenLifetime`',
+      );
+    }
+
+    if (!options.model) {
+      throw new InvalidArgumentError('Missing parameter: `model`');
+    }
+
+    if (!options.refreshTokenLifetime) {
+      throw new InvalidArgumentError(
+        'Missing parameter: `refreshTokenLifetime`',
+      );
+    }
+
+    if (!options.model.getClient) {
+      throw new InvalidArgumentError(
+        'Invalid argument: model does not implement `getClient()`',
+      );
+    }
+
+    this.accessTokenLifetime = options.accessTokenLifetime;
+    this.grantTypes = { ...grantTypes, ...options.extendedGrantTypes };
+    this.model = options.model;
+    this.refreshTokenLifetime = options.refreshTokenLifetime;
+    this.allowExtendedTokenAttributes = options.allowExtendedTokenAttributes;
+    this.requireClientAuthentication =
+      options.requireClientAuthentication || {};
+    this.alwaysIssueNewRefreshToken =
+      options.alwaysIssueNewRefreshToken !== false;
+  }
+
+  /**
+   * Token Handler.
+   */
+
+  async handle(request: Request, response: Response) {
+    if (!(request instanceof Request)) {
+      throw new InvalidArgumentError(
+        'Invalid argument: `request` must be an instance of Request',
+      );
+    }
+
+    if (!(response instanceof Response)) {
+      throw new InvalidArgumentError(
+        'Invalid argument: `response` must be an instance of Response',
+      );
+    }
+
+    if (request.method !== 'POST') {
+      throw new InvalidRequestError('Invalid request: method must be POST');
+    }
+
+    if (!request.is('application/x-www-form-urlencoded')) {
+      throw new InvalidRequestError(
+        'Invalid request: content must be application/x-www-form-urlencoded',
+      );
+    }
+
+    // Extend model object with request
+    this.model.request = request;
+
+    try {
+      const client = await this.getClient(request, response);
+      const data = await this.handleGrantType(request, client);
```

Comment:
The might resolve faster in Parallel.

Reply:
Do we need to add bluebird to this project to just use Bluebird.map, so that it is 1 ms faster?

Reply:
No, as client is a parameter for handleGrantType